openpype/lib/applications.py

```diff
@@ -1657,28 +1653,42 @@ def prepare_context_environments(data, env_group=None, modules_manager=None):
     system_settings = get_system_settings()
     data["project_settings"] = project_settings
     data["system_settings"] = system_settings
-    # Apply project specific environments on current env value
-    apply_project_environments_value(
-        project_name, data["env"], project_settings, env_group
-    )
 
     app = data["app"]
     context_env = {
         "AVALON_PROJECT": project_doc["name"],
-        "AVALON_ASSET": asset_doc["name"],
-        "AVALON_TASK": task_name,
         "AVALON_APP_NAME": app.full_name
     }
+    if asset_doc:
+        context_env["AVALON_ASSET"] = asset_doc["name"]
+        if task_name:
+            context_env["AVALON_TASK"] = task_name
 
     log.debug(
         "Context environments set:\n{}".format(
             json.dumps(context_env, indent=4)
         )
     )
     data["env"].update(context_env)
+
+    # Apply project specific environments on current env value
+    # - apply them once the context environments are set
+    apply_project_environments_value(
+        project_name, data["env"], project_settings, env_group
+    )
+
     if not app.is_host:
         return
 
+    data["env"]["AVALON_APP"] = app.host_name
+
+    if not not asset_doc or not task_name:
+        # QUESTION replace with log.info and skip workfile discovery?
+        # - technically it should be possible to launch host without context
```

Comment:
Does this affect the discoverability of the launcher? It would be cool to still allow people to launch hosts without a proper context

Reply:
> Does this affect the discoverability of the launcher? 

No. Launcher is a different story...